### PR TITLE
Fix #3848: Classify ArraySelect as side effecting

### LINKED ISSUE #3848: FunctionEmitter incorrectly classifies ArraySelect as side-effect free
Discovered while working on #3824.

https://github.com/scala-js/scala-js/blob/dacf08e14b0a856b89204d63d2b9450fa55a7e65/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala#L1147-L1148

This is incorrect, as it can throw an `ArrayIndexOutOfBoundsException`.

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala [hunk 1]
```diff
@@ -45,4 +45,15 @@ class ArrayTest {
     assertThrows(classOf[ArrayIndexOutOfBoundsException], a(Int.MaxValue) = 1)
   }
 
+  @Test
+  def arraySelectSideEffecting_issue_3848(): Unit = {
+      assumeTrue("Assuming compliant ArrayIndexOutOfBounds",
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala]:
```suggestion
    assumeTrue("Assuming compliant ArrayIndexOutOfBounds",
```

**Comment by gzm0** [comment]
Updated.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala [hunk 2]
```diff
@@ -45,4 +45,15 @@ class ArrayTest {
     assertThrows(classOf[ArrayIndexOutOfBoundsException], a(Int.MaxValue) = 1)
   }
 
+  @Test
+  def arraySelectSideEffecting_issue_3848(): Unit = {
+    assumeTrue("Assuming compliant ArrayIndexOutOfBounds",
+      hasCompliantArrayIndexOutOfBounds)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala]: Ah! The second line was supposed to stay where it was, to have a 4-space indentation :p

Comment (gzm0) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/ArrayTest.scala]: Oh dear :P

**Comment by gzm0** [comment]
Updated

**Pull request merged**